### ISSUE TITLE
Additional use cases for custom (processing) routine: Searches

### ISSUE THREAD

**Issue by tschoening** [kaitai-io/kaitai_struct#410]
We have [custom processing routines](http://doc.kaitai.io/user_guide.html#custom-process) now and in #51 more general callbacks/plugins/whatever have been discussed as well. Therefore I would like to at least document what I'm dealing with currently, because that use case in my opinion might be a good enhancement to what custom processing routines might be capable of in future. If you think otherwise, feel free to close or move/merge it to some more appropriate issue. Thanks!

I'm dealing with data consisting of multiple layers, pretty much like IP -> TCP -> HTTP -> SOAP -> user data, while my layers/protocols are actually different of course. Additionally, not all of those layers need to be present as well, at least one half of them is optional and only used by some manufacturers or within manufacturers only by some of their devices for various reasons.

The user data in this example is generated by devices which are known to sometimes have firmware bugs and therefore create wrong data, but if one knows which device generated the data, one is able to fix at least some of those problems. This is sometimes even possible directly in KS e.g. using default processes like `xor`.

Devices are identified by e.g. serial numbers, but devices can forward user data on behalf of other devices. So within all of those (optional) layers, I might have different serial numbers for different devices and need to find and decide which one is the correct one in some scenario. In most cases it's simply the most current/deepest serial number within the hierarchy of layers, but remember that those might be optional, so cannot be simply chained together and queried for such numbers.

In the end this is a recursive search through all the layers with handling which layer is actually available, checking which does provide a more detailed serial number and if not, defaulting to some always available fallback. My current implementation for what I need to support now in a dedicated search class in Java is ~200 lines of text, including some docs, so it maybe sounds more difficult than it actually is. :-) But I'm pretty sure to not support all possible layers and combinations yet as well.

So after having the implementation in Java I thought of giving an implementation in KS a try and came up with the following:

    instances:
      find_dll_lid_in_ell_with:
        value:  |
                control_info.is_ell_01    ? layer.as<fmt_oms_ell_01_with_next_layer>.next_layer.find_dll_lid
                                          : layer.as<fmt_oms_ell_03_with_next_layer>.next_layer.find_dll_lid
        if:     control_info.is_ell_01 or
                control_info.is_ell_03

      find_dll_lid_in_ell_opt:
        value:  |
                control_info.is_ell_02    ? layer.as<fmt_oms_ell_02_opt_next_layer>.next_layer_if.find_dll_lid
                                          : layer.as<fmt_oms_ell_04_opt_next_layer>.next_layer_if.find_dll_lid
        if:     (control_info.is_ell_02 and not layer.as<fmt_oms_ell_02_opt_next_layer>.ell_02.is_encrypted) or
                (control_info.is_ell_04 and not layer.as<fmt_oms_ell_04_opt_next_layer>.ell_04.is_encrypted)

      find_dll_lid_in_ell_with_or_opt:
        value:  |
                (control_info.is_ell_01 or control_info.is_ell_03)  ? find_dll_lid_in_ell_with
                                                                    : find_dll_lid_in_ell_opt

      find_dll_lid_in_ell:
        value:  |
                find_dll_lid_in_ell_with_or_opt ? find_dll_lid_in_ell_with_or_opt :
                  control_info.is_ell_03        ? layer.as<fmt_oms_ell_03_with_next_layer>.ell_03.lid
                                                : layer.as<fmt_oms_ell_04_with_next_layer>.ell_04.ell_03.lid
        if:     find_dll_lid_in_ell_with_or_opt or control_info.is_ell_03 or control_info.is_ell_04

      find_dll_lid_in_tpl_long_with:
        value:  |
                control_info.value        == 0x6B ? layer.as<fmt_oms_tpl_long_with_apl_format>.tpl.lid  :
                  control_info.value      == 0x6F ? layer.as<fmt_oms_tpl_long_with_apl_error>.tpl.lid   :
                    control_info.value    == 0x72 ? layer.as<fmt_oms_tpl_long_with_apl_full>.tpl.lid    :
                                                    layer.as<fmt_oms_tpl_long_with_apl_compact>.tpl.lid
        if:     control_info.is_tpl_long

      find_dll_lid_in_ell_or_tpl_long:
        value:  |
                control_info.is_ell ? find_dll_lid_in_ell
                                    : find_dll_lid_in_tpl_long_with
        if:     control_info.is_ell or control_info.is_tpl_long

      find_dll_lid:
        value:  |
                control_info.is_ell or control_info.is_tpl_long ? find_dll_lid_in_ell_or_tpl_long
                                                                : _root.lid

This doesn't compile properly currently, some type inferring problem somewhere, but is only a PoC anyway. Things like error handling, checking for available serial numbers, where I simply work with `null`  in Java, etc. seems pretty complicated to me. So in the end, I don't think that I would like to implement searches like these in KS directly, especially if one needs different searches in the end as well, which I do. In Java I simply create individual classes per scenario and maybe reuse static methods or whatever, this seems more complicated in KS.

Something like the following instead seems far more maintainable to me:

    instances:
      find_dll_lid:
        process:  find_dll_lid_impl(this vs. (control_info, layer))
        type:     fmt_oms_dll_lid

For my searching example, one would "only" need pretty much what custom processing routines are capable of today, only with different input and output. Like with value instances, the results should be cached and `type` is needed because from my understanding KSC can't know the result otherwise.

Any opinions? Is this something KS should/wants to care about in future?

**Comment by arekbulski** [comment]
TL;DR. You lost me there, buddy. :confused:

**Comment by tschoening** [comment]
I've added a summary.

**Comment by arekbulski** [comment]
I am still not able to understand the specifics of this feature, but that is my own shortcoming. For the record, it does somewhat remind me of a feature that was added in Construct a long time ago, although I wasnt the one who implemented it: 
https://github.com/construct/construct/blob/04ec2c1769c628bc8522609f434dff94ebc2d08e/construct/lib/containers.py#L265-L300

**Comment by GreyCat** [comment]
Just for the record: issues that might be related to this one:

* https://github.com/kaitai-io/kaitai_struct/issues/158
* https://github.com/kaitai-io/kaitai_struct/issues/530
* https://github.com/kaitai-io/kaitai_struct/issues/532
